graylog2-server/src/main/java/org/graylog2/inputs/transports/netty/HttpForwardedForHandler.java

```diff
@@ -0,0 +1,227 @@
+
+import com.google.common.base.Splitter;
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
+import com.google.common.net.InetAddresses;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpRequest;
+import org.graylog2.utilities.IpSubnet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class HttpForwardedForHandler extends SimpleChannelInboundHandler<HttpRequest> {
+    private static final Logger LOG = LoggerFactory.getLogger(HttpForwardedForHandler.class);
+
+    // Splits comma-separated forwarded-elements
+    private static final Splitter ELEMENT_SPLITTER =
+            Splitter.on(',')
+                    .omitEmptyStrings()
+                    .trimResults();
+
+    // Splits semicolon-separated parameters within an element
+    private static final Splitter PARAM_SPLITTER =
+            Splitter.on(';')
+                    .omitEmptyStrings()
+                    .trimResults();
+
+    // Splits key=value pairs
+    private static final Splitter KEYVALUE_SPLITTER =
+            Splitter.on('=')
+                    .limit(2)
+                    .trimResults();
+
+    private final boolean enableForwardedFor;
+    private final boolean enableRealIpHeader;
+    private final List<String> realIpHeaders;
+    private final boolean requireTrustedProxies;
+    private final Set<IpSubnet> trustedProxies;
+
+    /**
+     * @param enableForwardedFor    true if we should parse X-Forwarded-For and Forwarded headers
+     * @param realIpHeaders         a comma separated list to check for original IP values, if null or empty don't attempt to find a value
+     * @param requireTrustedProxies true if all other values in the chain are checked against our set of trusted proxies
+     * @param trustedProxies        config setting from server conf, the subnets of proxies we accept forwarded-for headers from
+     */
+    public HttpForwardedForHandler(boolean enableForwardedFor, boolean enableRealIpHeader, String realIpHeaders, boolean requireTrustedProxies, Set<IpSubnet> trustedProxies) {
+        this.enableForwardedFor = enableForwardedFor;
+        this.enableRealIpHeader = enableRealIpHeader;
+        this.realIpHeaders = Splitter.on(',').omitEmptyStrings().trimResults().splitToList(Strings.nullToEmpty(realIpHeaders));
+        this.requireTrustedProxies = requireTrustedProxies;
+        this.trustedProxies = trustedProxies;
+    }
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx, HttpRequest msg) throws Exception {
```

Comment:
I see the following error message in my server logs:

```
2025-06-19 15:07:34,597 ERROR: org.graylog2.plugin.inputs.transports.AbstractTcpTransport - Error in Input [Raw HTTP/HaTeTePe/685414bd6d707bb319eea75e] (channel [id: 0x460ec98d, L:/172.18.0.14:5557 ! R:/172.18.0.5:44242]) (cause io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1)
```

This sounds like a missing `.release()` call in the channel handler.

Reply:
added a retain and switch everything to using `FullHttpRequest` to make it more clear